### PR TITLE
[MRG] Ignore NaNs when averaging spectrograms

### LINKED ISSUE #7024: Error in raw.plot_psd() with bad segments
I think #6748 broke PSD plotting when bad segments are present in annotations. The reason is that `psd_welch` now averages PSDs without taking care of NaNs, which it should because bad annotations are automatically transformed into `np.nan`.

```
import mne
from mne.io import read_raw_fif
from mne.datasets import sample
from mne.time_frequency import psd_welch


path = sample.data_path()
raw = read_raw_fif(path + "/MEG/sample/sample_audvis_filt-0-40_raw.fif")
picks = mne.pick_types(raw.info, meg=False, eeg=True)
raw.set_annotations(mne.Annotations(0, 10, "bad_segment"))
psds, freqs = psd_welch(raw, picks=picks)
```

In this example, all elements in `psds` are `np.nan`. Previously, when only mean was supported, we used `np.nanmean` to ignore NaNs. Not sure if there is an easy fix because `scipy.signal.welch` does not support NaNs AFAIK.

### PR DESCRIPTION
**Pull request by cbrnr**
Fixes #7024.

If this works, I'll add a test.

### CODE REVIEW

**Comment by cbrnr** [comment]
Turns out that we already had a test for NaNs in the data, but it was broken. I fixed it and I think this is ready for merge.

**Comment by cbrnr** [comment]
@agramfort @larsoner any chance we could backport this and make a bugfix release soon?

File: mne/time_frequency/psd.py [hunk 1]
```diff
@@ -124,6 +124,10 @@ def psd_array_welch(x, sfreq, fmin=0, fmax=np.inf, n_fft=256, n_overlap=0,
     -----
     .. versionadded:: 0.14.0
     """
+    if average is not None and average not in ["mean", "median"]:
+        raise ValueError('average must be one of `mean`, `median`, or None, '
+                         'got {}'.format(average))
```
Comment (larsoner) [inline mne/time_frequency/psd.py]: `_check_option('average', average, (None, 'mean', 'median'))`

File: mne/time_frequency/psd.py [hunk 2]
```diff
@@ -140,34 +144,20 @@ def psd_array_welch(x, sfreq, fmin=0, fmax=np.inf, n_fft=256, n_overlap=0,
...
+                                     n_per_seg=n_per_seg, nfft=n_fft, fs=sfreq,
+                                     freq_mask=freq_mask, func=spectrogram)
+                       for d in x_splits)
+    psds = np.concatenate(f_spect, axis=0)
+    psds.shape = dshape + (len(freqs), -1)
 
+    if average == 'mean':
+        psds = np.nanmean(psds, axis=-1)
+    elif average == 'median':
+        psds = np.nanmedian(psds, axis=-1)
```
Comment (larsoner) [inline mne/time_frequency/psd.py]: This is going to have a much higher (by a factor of `n_epochs`) memory cost than before. If you write a little `_welch_func` wrapper that does the mean and median (if requested) internally, then loop over that, the cost will go back down.

Comment (cbrnr) [reply mne/time_frequency/psd.py]: I'm not sure I understand since that's exactly how we did it before we changed to `scipy.signal.welch`. Do you want me to compute a rolling mean/median in a loop?

File: mne/time_frequency/psd.py (hunk 1 above)
Comment (cbrnr) [reply mne/time_frequency/psd.py]: Done

File: mne/time_frequency/psd.py (hunk 2 above)
Comment (larsoner) [reply mne/time_frequency/psd.py]: Pushed a commit, hopefully it makes it clearer

File: mne/time_frequency/psd.py [hunk 3]
```diff
@@ -6,28 +6,22 @@
...
-def _welch_func(epoch, n_overlap, n_per_seg, nfft, fs, freq_mask, average,
-                func):
-    """Aux function."""
-    kws = dict(fs=fs, nperseg=n_per_seg, noverlap=n_overlap, nfft=nfft,
-               window='hamming', average=average)
-
-    if average == 'mean':  # Compatibility with SciPy <1.2
-        del kws['average']
-
-    _, psd = func(epoch, **kws)
```
Comment (larsoner) [inline mne/time_frequency/psd.py]: ... previously this func was `welch`, so within each `_welch_func` call the average across windows was taken. My commit preserves this behavior to save memory.

Comment (cbrnr) [reply mne/time_frequency/psd.py]: But isn't the function parallelized over channels? And by default `n_jobs=1` so nothing runs in parallel.

Comment (larsoner) [reply mne/time_frequency/psd.py]: Whatever the parallelization is over, computing and averaging each within the loop then concatenating should consume less memory than computing each within a loop, concatenating, then averaging. Make sense?

Comment (cbrnr) [reply mne/time_frequency/psd.py]: You mean potentially, right? Then yes, this makes sense. For `n_jobs=1` there shouldn't be a difference though.

Comment (larsoner) [reply mne/time_frequency/psd.py]: Even for one job the memory should be lower

Comment (larsoner) [reply mne/time_frequency/psd.py]: Before this PR (and after my commit in this PR), the operations (when `average is not None`) were:

1. For each channel/signal/whatever (parallelized or not):
    - Compute STFT to get something of shape `(n_freqs, n_segments)`
    - Average over windows/segments (these two steps were implicitly done by `welch` and in my commit are explicitly coded)
2. Concatenate results that are shape `(..., n_freqs)` to form the output

After your changes on top of `master` (before I pushed my commit) the operations were:

1. For each channel/signal/whatever (parallelized or not):
    - Compute STFT to get something of shape `(n_freqs, n_segments)`
2. Concatenate results that are shape `(..., n_freqs, n_segments`): requires memory of size `(..., n_freqs, n_segments)`, i.e. `n_segments` times more memory than step (2) above
3. Average over segments to obtain a result of shape `(..., n_freqs)` to form the output

Make sense?

Comment (cbrnr) [reply mne/time_frequency/psd.py]: Maybe it's because it's already late, but no, I don't get it. Let's look at the default case of not parallelized, so `n_jobs=1` and `len(x_splits) = 1`. Then after your latest commit:
1. Compute spectrograms of shape `(n_channels, n_freqs, n_segments)`.
2. Average over segments to get shape `(n_channels, n_freqs)`.
3. Concatenate results, which is a no-op because `len(f_spect) = 1`.

My version:
1. Compute spectrograms of shape `(n_channels, n_freqs, n_segments)`.
2. Concatenate results, which is a no-op because `len(f_spect) = 1`.
3. Average over segments to get shape `(n_channels, n_freqs)`.

So in both versions we go from `(n_channels, n_freqs, n_segments)` to `(n_channels, n_freqs)` with basically the same steps because we're not concatenating anything.

Comment (larsoner) [reply mne/time_frequency/psd.py]: Ahh you're right, it will only matter in the parallelized case currently. I hadn't noticed that the parallelization was over an `array_split` of X and not just X itself. So I was wrong about the dimensions here -- good catch!

However, `np.concatenate(...)` always makes a copy, so even if it's just a list with a single element it's not a no-op, it copies. So there still will be a factor of 2x memory pain from not doing it inside the loop -- but at least it's much less than I thought.

FWIW we should actually probably change the parallelization behavior since, unlike in some of our functions, there won't be any benefit here of iterating over the `array_split`, and there is a (potentially quite large) memory cost to doing it this way.

Feel free to rebase and then this should be good to go. I say we keep my version so that we at least get the 2x (plus more in parallelization) benefit in the current case.

Comment (cbrnr) [reply mne/time_frequency/psd.py]: > Ahh you're right, it will only matter in the parallelized case currently. I hadn't noticed that the parallelization was over an `array_split` of X and not just X itself. So I was wrong about the dimensions here -- good catch!

😌 - took me some time to figure it out and I'm still not sure I got it all sorted out (see below).

> However, `np.concatenate(...)` always makes a copy, so even if it's just a list with a single element it's not a no-op, it copies. So there still will be a factor of 2x memory pain from not doing it inside the loop -- but at least it's much less than I thought.

I think it's actually not just 2x, but `n_segments`x. In your version, we concatenate (make a copy of) `[(n_channels, n_freqs)]` and in my version we concatenate (make a copy of) `[(n_channels, n_freqs, n_segments)]`. So we should definitely keep your version.

> FWIW we should actually probably change the parallelization behavior since, unlike in some of our functions, there won't be any benefit here of iterating over the `array_split`, and there is a (potentially quite large) memory cost to doing it this way.

I was wondering why we need the parallelization at all here. `scipy.signal.spectrogram` is already vectorized over channels/all dimensions except the one we compute over (I assume), so what is the use case for splitting up? And what is split up?

**Comment by cbrnr** [comment]
I've rebased so this could be merged. However, I'd still like to clarify whether or not parallelization makes sense here, because IMO `scipy.signal.spectrogram` should take care of vectorizing and therefore we don't need to parallelize ourselves. This makes the code much easier to read too.

**Comment by cbrnr** [comment]
In fact, using the test data from #7024, it takes around 0.49s for `n_jobs=1` and around 2.00s for `n_jobs>=2`.

**Comment by agramfort** [comment]
I would not be surprised if n_jobs > 1 is slower

>

**Comment by cbrnr** [comment]
I'm not surprised that it is slower either, but I wonder why this is still in the function. Unless there is a use case that I'm totally missing right now, I'd say we remove it completely.

**Comment by larsoner** [comment]
Even code that is vectorized can use only a single core on a multi-core machine. Using one vs multiple cores in vectorized code depends on at least:

1. the linear algebra / FFT libraries used under the hood by NumPy and SciPy
2. the options for those libraries (e.g., OMP_NUM_THREADS), and
3. the actual functions being called

If the bulk of the time is spent in multithreaded BLAS or FFT calls, and the user has a multithreading BLAS/FFT library, and they have multithreading enabled for that library, then indeed there shouldn't be much benefit to parallelization (and even some loss possibly). If these conditions aren't satisfied, then there can be. Hence why we would probably still allow parallelization in cases like this, it can depend on the user's setup.

We can investigate whether or not there is a benefit under different circumstances / setups and evaluate if there is any benefit to `n_jobs` here, but we should probably do that separately.

**Comment by cbrnr** [comment]
Agreed, we can do this separately. There's no performance drop when keeping our parallelization and using `n_jobs=1`, so it is not a pressing issue. I think most people should be running a multithreaded linear algebra library nowadays (at least this is the case for Anaconda on all major platforms, and Python on macOS uses the Accelerate framework), so most people will likely see a dramatic performance drop (in my setup 1/4x) when using more than 1 job.

In summary, this is ready for merge.

**Comment by larsoner** [comment]
Thanks @cbrnr

**Comment by larsoner** [comment]
Backported. @agramfort should we just push out a quick 0.19.2? I think we can just do a quick `twine` call
